### PR TITLE
refactor(replicator): listener in Replicator struct

### PR DESCRIPTION
**Pull request by adrien-jeser-doctolib**

### CODE REVIEW

File: build.rs [hunk 1]
```diff
@@ -41,13 +41,12 @@ fn main() -> Result<(), Box<dyn Error>> {
 
     // Bypass copying libraries when the build script is called in a cargo check context.
     if env::var("ONLY_CARGO_CHECK").unwrap_or_default() != *"true" {
-        copy_lib().expect(
-            &format!(
+        copy_lib().unwrap_or_else(|_| {
+            panic!(
                 "can't copy cblite libs, is '{}' a supported target?",
-                env::var("TARGET")?
+                env::var("TARGET").unwrap_or_default()
```
Comment (n-eq) [inline build.rs:47]: In what case is this used?

**Review by n-eq — APPROVED** [review]

File: src/replicator.rs [hunk 2]
```diff
@@ -487,12 +487,16 @@ pub struct ReplicatorConfiguration {
...
 
+type ReplicatorsListeners<T> = Vec<Listener<Box<T>>>;
+
 /** A background task that syncs a \ref Database with a remote server or peer. */
 pub struct Replicator {
     cbl_ref: *mut CBLReplicator,
     pub config: Option<ReplicatorConfiguration>,
     pub headers: Option<MutableDict>,
     pub context: Option<Box<ReplicationConfigurationContext>>,
+    change_listener: ReplicatorsListeners<ReplicatorChangeListener>,
```
Comment (MatthieuClochardDoctolib) [inline src/replicator.rs:498]:
```suggestion
    change_listeners: ReplicatorsListeners<ReplicatorChangeListener>,
```

File: src/replicator.rs [hunk 3]
```diff
@@ -487,12 +487,16 @@ pub struct ReplicatorConfiguration {
...
+    document_listener: ReplicatorsListeners<ReplicatedDocumentListener>,
```
Comment (MatthieuClochardDoctolib) [inline src/replicator.rs:499]:
```suggestion
    document_listeners: ReplicatorsListeners<ReplicatedDocumentListener>,
```

File: src/replicator.rs [hunk 4]
```diff
@@ -570,6 +574,8 @@ impl Replicator {
                 config: Some(config),
                 headers: Some(headers),
                 context: Some(context),
+                change_listener: vec![],
```
Comment (MatthieuClochardDoctolib) [inline src/replicator.rs:577]:
```suggestion
                change_listeners: vec![],
```

File: src/replicator.rs [hunk 5]
```diff
@@ -570,6 +574,8 @@ impl Replicator {
...
+                document_listener: vec![],
```
Comment (MatthieuClochardDoctolib) [inline src/replicator.rs:578]:
```suggestion
                document_listeners: vec![],
```

**Review by MatthieuClochardDoctolib — APPROVED** [review]

File: src/replicator.rs [hunk 6]
```diff
@@ -627,6 +633,89 @@ impl Replicator {
...
+                )),
+                Box::from_raw(ptr),
+            )
+        };
+        self.change_listener.push(listener);
+        self
+    }
+
+    /** Adds a listener that will be called when documents are replicated. */
+    #[must_use]
```
Comment (MatthieuClochardDoctolib) [inline src/replicator.rs:701]: Should it be must_use ?

Comment (adrien-jeser-doctolib) [reply src/replicator.rs:701]: Yes

Comment (adrien-jeser-doctolib) [reply src/replicator.rs:701]: If the `Replicator` structure is not used then its lifetime necessarily stops at the end of the scope, so the replication stops

**Pull request merged**